sonar-maven-plugin/src/main/java/org/sonarsource/scanner/maven/bootstrap/Maven3ToolchainResolver.java

```diff
@@ -0,0 +1,122 @@
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.nio.file.Path;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.apache.maven.execution.MavenSession;
+import org.apache.maven.plugin.MojoExecution;
+import org.apache.maven.plugin.logging.Log;
+import org.apache.maven.toolchain.Toolchain;
+import org.apache.maven.toolchain.ToolchainManager;
+import org.apache.maven.toolchain.java.DefaultJavaToolChain;
+import org.apache.maven.toolchain.java.JavaToolchain;
+import org.codehaus.plexus.configuration.PlexusConfiguration;
+import org.codehaus.plexus.configuration.xml.XmlPlexusConfiguration;
+import org.codehaus.plexus.util.xml.Xpp3Dom;
+
+import static org.apache.commons.lang3.StringUtils.isNotEmpty;
+import static org.sonarsource.scanner.maven.bootstrap.MavenCompilerResolver.toJdkHomeFromJavacExec;
+import static org.sonarsource.scanner.maven.bootstrap.MavenUtils.convertString;
+
+public class Maven3ToolchainResolver implements ToolchainResolver {
+
+  private final MavenSession session;
+  private final Log log;
+  private final ToolchainManager toolchainManager;
+
+  public Maven3ToolchainResolver(MavenSession session, Log log, ToolchainManager toolchainManager) {
+    this.session = session;
+    this.log = log;
+    this.toolchainManager = toolchainManager;
+  }
+
+  @Override
+  public Optional<Path> getJdkHomeFromToolchains(MojoExecution compilerExecution) {
+
+    // Inspired by
+    // https://github.com/apache/maven-compiler-plugin/blob/dc4a5635ba4eb2ba5e461fa53b2c47c58d7fa397/src/main/java/org/apache/maven/plugin/compiler/AbstractCompilerMojo.java#L1418
+    Toolchain tc = null;
+
+    // Maven 3.3.1 has plugin execution scoped Toolchain Support
+    Optional<Map<String, String>> jdkToolchain = getMapConfiguration(compilerExecution, "jdkToolchain");
+    if (jdkToolchain.isPresent() && !jdkToolchain.get().isEmpty()) {
+      List<Toolchain> tcs = collectMatchingToolchains(jdkToolchain.get());
+      if (!tcs.isEmpty()) {
+        tc = tcs.get(0);
+      }
+    }
+
+    // Fallback on the global jdk toolchain
+    if (tc == null) {
+      tc = toolchainManager.getToolchainFromBuildContext("jdk", session);
+    }
+
+    if (tc != null) {
+      // DefaultJavaToolChain is the older API, supported since Maven 3.2.3, but removed in Maven 4, so let's check it first
+      if (tc instanceof JavaToolchain) {
+        DefaultJavaToolChain defaultJavaToolChain = (DefaultJavaToolChain) tc;
+        String jdkHome = defaultJavaToolChain.getJavaHome();
+        if (isNotEmpty(jdkHome)) {
+          return Optional.of(Path.of(jdkHome).toAbsolutePath());
+        }
+      }
+
+      // Fallback on locating the "javac" executable, this should be working as well with Maven 4. Alternative would be to cast tc to
+      String javacToUse = tc.findTool("javac");
```

Comment:
We seem to be defaulting to the home of runtime `javac` in the invoker tests running with Maven 4. From my debugging earlier this week, we might want to at tc as a [DefaultToolChainManagerV3](https://maven.apache.org/ref/4.0.0-rc-4/maven-compat-modules/maven-compat/xref/org/apache/maven/toolchain/ToolchainManagerFactory.html#L148) from which we also have to call `getToolChains`. Let's have a look at it together.